proposals/0000-template.md

```diff
@@ -0,0 +1,154 @@
+    /// Threshold for closing the poll, i.e, if the ratio of stake on a certain proposal over total stake reaches
+    /// threshold, the poll is closed.
+    threshold: Fraction,
+    /// Voting result. `None` means the poll is still open.
+    result: Option<ProposalId>,
+}
+```
+
+On the top level we maintain a poll where there are multiple proposals that can be voted on. The poll keeps track of the
+proposals and accounts that have participated in the poll so far and their total voted stake. When the stake voted on a given
+proposal reaches the threshold for this poll, the proposal becomes final and the poll is closed.
+
+As an important constituent of the poll, `Proposal` contains all the voting information of this particular proposal,
+in addition to the actual content of the proposal:
+```rust
+pub struct Proposal {
+    /// Human readable description of the proposal.
+    description: String,
+    /// Serialized metadata of the proposal.
+    metadata: String,
+    /// When this proposal expires.
+    expiration_height: BlockHeight,
+    /// Current votes on this proposal.
+    votes: Map<AccountId, Balance>,    
+}
+```
+Here `metadata` is the json serialized content of the proposal. For example, if the poll is about when the network upgrade
+should happen, `metadata` will be something like `{"height": 10000}`. Validators can vote on multiple different proposals
+at the same time, provided that the sum of stake voted does not exceed their current stake.
+
+The voting contract has the following methods to allow creation of polls and proposals, as well as voting for proposals.
+
+```rust
+impl Poll {
+    /// Initialize the poll on some topic.
+    pub fn new(topic: String, threshold: Fraction) -> Self;
+    /// Create a proposal for a given poll.
+    pub fn create_proposal(&mut self, description: String, metadata: String);
+    /// Vote on a given proposal with certain amont of stake.
+    pub fn vote(&mut self, proposal_id: ProposalId, stake: Balance);
+    /// View proposal
+    pub fn get_proposal(&mut self, proposal_id: ProposalId) -> Proposal;
+}
+```
+Notice that the `vote` function can also be used to withdraw a vote by putting 0 stake on the vote.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A major difficulty in managing the votes comes from the fact that validator stake can change from epoch to epoch, which
+means that we need to carefully update validator votes. For example, if a validator has 100 stake in the previous epoch and
+they voted 80 on some proposal, but in the current epoch their stake has decreased to 80, and now if they try to vote 20
+on some other proposal, it should fail. To address this, we introduce a helper function `resolve`, which is called
+every epoch before any call to the contract:
+```rust
+fn resolve(&mut self) {
+    // go through all the proposals and for each account, scale the stake on votes to current stake, i.e,
```

Comment:
This is not going to work. Need to pass the `proposal_id`. Otherwise you'll exceed the gas limit.

Reply:
Make sense. But how do we make sure that we resolve all proposals?

Reply:
If one of the proposals is resolved, then the entire poll is resolved.

Reply:
That doesn't make sense. Here `resolve` scales the stake voted on proposals. Resolving one proposal doesn't help.